test/test_upload_processor.py

```diff
@@ -0,0 +1,155 @@
+""" This file tests the functions defined in upload_processor """
+
+import unittest
+import subprocess
+import os
+from typing import List
+import hailtop.batch as hb
+from joint_calling.upload_processor import batch_move_files
+
+
+def validate_move(upload_prefix: str, main_prefix: str, sample: str) -> bool:
+    """Checks that a given file exists at the location in
+    the main bucket and no longer exists in the upload bucket.
+    Returns True if this is the case and False otherwise.
+    """
+
+    upload_path = os.path.join('gs://', upload_prefix, sample)
+    main_path = os.path.join('gs://', main_prefix, sample)
+
+    exists_main = subprocess.run(['gsutil', '-q', 'stat', main_path], check=False)
+    exists_upload = subprocess.run(['gsutil', '-q', 'stat', upload_path], check=False)
+
+    # Exists at destination and not at source
+    if exists_upload.returncode == 1 and exists_main.returncode == 0:
+        return True
+
+    return False
+
+
+def cleanup(bucket: str, samples: List[str]):
+    """Clean up function that deletes remaining files after
+    test run"""
+
+    for sample in samples:
+        full_path = os.path.join('gs://', bucket, sample)
+        subprocess.run(['gsutil', 'rm', full_path], check=False)
+
+
+def upload_files(files: List[str], upload_prefix: str):
+    """A function to mimic file upload. Takes a list of
+    file names, creates these files, then moved them into
+    the upload bucket used for further testing."""
+
+    for f in files:
+        subprocess.run(['touch', f], check=True)
+        full_path = os.path.join('gs://', upload_prefix, f)
+        subprocess.run(['gsutil', 'mv', f, full_path], check=True)
+
+
+class TestUploadProcessor(unittest.TestCase):
+    """Test cases for the upload processor"""
+
+    def setUp(self):
+        """ Initialises standard variables for testing """
+        self.upload_prefix = 'cpg-fewgenomes-temporary/vivian-test/test-upload'
+        self.main_prefix = 'cpg-fewgenomes-temporary/vivian-test/test-main'
+        self.docker_image = os.environ.get('DOCKER_IMAGE')
+        self.key = os.environ.get('GSA_KEY')
+
+    def test_batch_move_standard(self):
+        """ Testing standard case of moving a list of files with valid inputs"""
+        sample_list = ['Sample5.gVCF', 'Sample6.gVCF', 'Sample7.gVCF']
+        upload_files(sample_list, self.upload_prefix)
+        batch = hb.Batch(name='Test Batch Move Standard')
+        batch_move_files(
+            batch,
+            sample_list,
+            self.upload_prefix,
+            self.main_prefix,
+            self.docker_image,
+            self.key,
+        )
+        batch.run()
+        # Check that the files have been moved to main
+        for sample in sample_list:
+            self.assertTrue(validate_move(self.upload_prefix, self.main_prefix, sample))
+
+        cleanup(self.main_prefix, sample_list)
```

Comment:
This won't get called if there's an exception during the test. Can we include this as part of the `tearDown` method?

Reply:
Good point! From what I can see the `tearDown` method is run at the end of each individual test. As such, I'm not quite sure of the best way to implement this. For example, we could just attempt to delete all the files (for each test) in the `tearDown` method. The tests will still run and pass but we would get a bunch of `CommandException: No URLs matched:` for the files that are used in different test cases. 

Alternatively, we could also just have one `samples` array that we use across all of the tests, this could be defined within `setUp` and then always deleted? 

Or there might be another approach I haven't considered. 

What do you think is the best approach?

Reply:
Good point! From what I can tell the `tearDown` method is called after each individual test is run. As such, I'm not sure of the best way to go about implementing this. 

For example, we could clean-up all of the test files in the `tearDown` method. This works and all the tests pass. But since we use different files for each test, there would be a bunch of `CommandException: No URLs matched:` for each file that wasn't used in the most recent one. 

Alternatively, we can simply use one standard set of files for all tests (perhaps defined in setUp), and then the `tearDown` method would be deleting the same files each time. 

There could also be another way that I haven't considered. 

What do you think would be the best approach?

Reply:
I'm assuming that the tests run sequentially (i.e. not concurrently), which I think is true for Python.

Could you create a random directory suffix (e.g. `test-13214a`) in `setUp`, use that directory for all files that are created during this "run", and delete the whole directory in `tearDown`? This way you could (1) still use whatever file names are convenient for a particular test and (2) potentially have different test runs be executed concurrently without interfering with each other.

Reply:
Great, thanks for the suggestion! I've implemented this in the latest commit :)